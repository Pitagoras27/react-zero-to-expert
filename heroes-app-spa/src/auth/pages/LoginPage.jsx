import { useNavigate } from "react-router-dom";

export const LoginPage = () => {
  const navigate = useNavigate();

  const onLogin = () => {
    // ? remember `replace` manage history stack of browser
    navigate("/marvel", { replace: true });
  };

  return (
    <div className="container m-3">
      <h1>Login</h1>
      <hr />
      <button className="btn btn-primary" onClick={onLogin}>
        Login
      </button>
    </div>
  );
};
