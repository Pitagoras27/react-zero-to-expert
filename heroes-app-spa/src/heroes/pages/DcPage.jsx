import { HeroList } from "../components";

export const DcPage = () => {
  return (
    <div className="container m-4">
      <h1>DC Page</h1>
      <HeroList publisher={"DC Comics"} />
    </div>
  );
};
