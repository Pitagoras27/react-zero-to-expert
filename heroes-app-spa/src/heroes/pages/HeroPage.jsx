import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom";
import { getHeroesById } from "../helpers";

export const HeroPage = () => {
  const { id } = useParams();
  const hero = useMemo(() => getHeroesById(id), [id]);
  const navigate = useNavigate();

  const onNavigateBack = () => {
    const lastPage = id.split("-")[0];
    navigate(`/${lastPage}`);
    // ? -1 params retuns last page of navigation
    // navigate("/marvel");
  };

  if (!hero) {
    return <Navigate to="/marvel" />;
  }

  return (
    <div className="container m-3">
      <div className="row">
        <div className="col-4">
          <img
            src={`/assets/heroes/${hero.id}.jpg`}
            alt={hero.superhero}
            className="img-thumbnail animate__animated animate__fadeInLeft animate__animated animate__bounceIn"
          />
        </div>
        <div className="col-8 animate__animated animate__bounceInRight">
          <h3 className="mb-3">{hero.superhero}</h3>
          <ul className="list-group list-group-flush">
            <li className="list-group-item">
              <span>Publisher: </span>
              {hero.publisher}
            </li>
            <li className="list-group-item">
              <span>Alter Ego: </span>
              {hero.alter_ego}
            </li>
            <li className="list-group-item">
              <span>First Appearance: </span>
              {hero.first_appearance}
            </li>
          </ul>
          <h5 className="mt-3">Characters</h5>
          <p>{hero.characters}</p>
          <hr />
          <button className="btn btn-primary" onClick={onNavigateBack}>
            Returns
          </button>
        </div>
      </div>
    </div>
  );
};
