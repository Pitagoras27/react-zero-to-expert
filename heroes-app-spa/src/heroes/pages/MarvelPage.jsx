import { HeroList } from "../components";

export const MarvelPage = () => {
  return (
    <div className="container m-4">
      <h1>Marvel Page</h1>

      <HeroList publisher={"Marvel Comics"} />
    </div>
  );
};
