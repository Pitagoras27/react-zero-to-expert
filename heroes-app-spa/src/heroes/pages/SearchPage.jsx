import queryString from "query-string";
import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { searching, onChange, onResetForm } = useForm({
    searching: "",
  });

  const onHandleSubmit = (e) => {
    e.preventDefault();
    onResetForm();
    console.log(q);
  };

  const { q = "a" } = queryString.parse(location.search);

  return (
    <div className="container m-3">
      <div className="row">
        <div className="col-md-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onHandleSubmit}>
            <input
              type="text"
              placeholder="search a hero"
              name="searching"
              autoComplete="off"
              className="form-control"
              value={searching}
              onChange={onChange}
            />
            <button type="submit" className="btn btn-outline-primary mt-3">
              Search
            </button>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />
          <div className="alert alert-info">Search a hero</div>
          <div className="alert alert-danger">Don't find hero!</div>
        </div>
      </div>
    </div>
  );
};
