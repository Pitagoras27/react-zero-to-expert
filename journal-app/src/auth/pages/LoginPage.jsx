import { Google } from "@mui/icons-material";
import {
  Alert,
  Button,
  Grid,
  Link,
  TextField,
  Typography,
} from "@mui/material";
import { useMemo } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link as RouterLink } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import {
  startLoginWithEmailPassword,
  startSigningWithGoogle,
} from "../../store/auth";
import { AuthLayout } from "../layout/AuthLayout";

export const LoginPage = () => {
  const { email, password, handleChange } = useForm({
    email: "",
    password: "",
  });
  const dispatch = useDispatch();
  const { status, errorMessage } = useSelector((state) => state.auth);
  const isAuthenticated = useMemo(() => status === "checking", [status]);

  const onSubmit = (e) => {
    e.preventDefault();
    // dispatch(checkingAuthentication());
    dispatch(startLoginWithEmailPassword({ email, password }));
  };

  const onGoogleOnSingIn = () => {
    dispatch(startSigningWithGoogle());
  };

  return (
    <AuthLayout title="Login">
      <form
        onSubmit={onSubmit}
        className="animate__animated animate__fadeIn animate__faster"
      >
        <Grid container>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Correo"
              name="email"
              type="email"
              value={email}
              placeholder="correo@google.com"
              autoComplete="off"
              onChange={handleChange}
              fullWidth
            />
          </Grid>
        </Grid>

        <Grid container spacing={2} sx={{ mb: 2 }}>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Password"
              name="password"
              type="password"
              value={password}
              placeholder="Password"
              autoComplete="off"
              onChange={handleChange}
              fullWidth
            />
          </Grid>
        </Grid>

        <Grid
          item
          xs={12}
          sx={{ mb: 2, mt: 1 }}
          display={!!errorMessage ? "" : "none"}
        >
          <Alert severity="error">{errorMessage}</Alert>
        </Grid>

        <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
          <Grid item xs={12} sm={6}>
            <Button
              type="submit"
              variant="contained"
              disabled={isAuthenticated}
              fullWidth
            >
              Login
            </Button>
          </Grid>
          <Grid item xs={12} sm={6}>
            <Button
              onClick={onGoogleOnSingIn}
              variant="contained"
              disabled={isAuthenticated}
              fullWidth
            >
              <Google />
              <Typography sx={{ ml: 1 }}>Google</Typography>
            </Button>
          </Grid>

          <Grid container direction="row" justifyContent="end">
            <Link component={RouterLink} color="inherit" to="/auth/register">
              <Typography sx={{ mt: 2 }}>Create Account</Typography>
            </Link>
          </Grid>
        </Grid>
      </form>
    </AuthLayout>
  );
};
