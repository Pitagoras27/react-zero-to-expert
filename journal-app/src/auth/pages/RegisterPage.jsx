import {
  Alert,
  Button,
  Grid,
  Link,
  TextField,
  Typography,
} from "@mui/material";
import { useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link as RouterLink } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { startCreatingUserWithEmailPassword } from "../../store/auth";
import { AuthLayout } from "../layout/AuthLayout";

const dataFormRegister = {
  displayName: "",
  email: "",
  password: "",
};

const validatedFields = {
  displayName: [
    (value, b) => !value.length >= 1,
    "The full name don't have empty",
  ],
  email: [(value) => !value.includes("@"), "The email is not valid"],
  password: [
    (value) => !(value.length >= 6),
    "The password must have 6 characters minimun",
  ],
};

export const RegisterPage = () => {
  const dispath = useDispatch();
  const { status, errorMessage } = useSelector((state) => state.auth);
  const {
    inputValue,
    displayName,
    email,
    password,
    handleChange,
    displayNameValid,
    emailValid,
    passwordValid,
    isFormValid,
  } = useForm(dataFormRegister, validatedFields);

  const isAuthenticatedChecking = useMemo(
    () => status === "checking",
    [status]
  );
  const [submitValidation, setSubmitValidation] = useState(false);
  console.log("errorMessage--->", errorMessage);
  const handleSubmit = (e) => {
    e.preventDefault();
    setSubmitValidation(true);

    if (!isFormValid) return;
    dispath(startCreatingUserWithEmailPassword(inputValue));
  };
  return (
    <AuthLayout title="Register">
      {/* <h4>The form is {isFormValid ? "valid" : "incorrect"}</h4> */}
      <form onSubmit={handleSubmit}>
        <Grid container>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Full Name"
              type="text"
              placeholder="Full Name"
              autoComplete="off"
              fullWidth
              value={displayName}
              onChange={handleChange}
              name="displayName"
              error={!!displayNameValid && submitValidation}
              helperText={displayNameValid}
            />
          </Grid>
        </Grid>

        <Grid container>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Correo"
              type="email"
              placeholder="correo@google.com"
              autoComplete="off"
              fullWidth
              value={email}
              onChange={handleChange}
              name="email"
              error={!!emailValid && submitValidation}
              helperText={emailValid}
            />
          </Grid>
        </Grid>

        <Grid container spacing={2} sx={{ mb: 2 }}>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Password"
              type="password"
              placeholder="Password"
              autoComplete="off"
              fullWidth
              value={password}
              onChange={handleChange}
              name="password"
              error={!!passwordValid && submitValidation}
              helperText={passwordValid}
            />
          </Grid>
        </Grid>

        <Grid item xs={12} display={!!errorMessage ? "" : "none"}>
          <Alert severity="error">{errorMessage}</Alert>
        </Grid>

        <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
          <Grid item xs={12} sm={6}>
            <Button
              variant="contained"
              fullWidth
              type="submit"
              disabled={isAuthenticatedChecking}
            >
              <Typography>Register</Typography>
            </Button>
          </Grid>

          <Grid container direction="row" justifyContent="end">
            <Link component={RouterLink} color="inherit" to="/auth/login">
              <Typography sx={{ mt: 2 }}>You have an Account</Typography>
            </Link>
          </Grid>
        </Grid>
      </form>
    </AuthLayout>
  );
};
