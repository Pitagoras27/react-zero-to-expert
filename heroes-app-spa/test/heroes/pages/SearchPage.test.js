import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockUseNavigate,
}));

describe("test over search component", () => {
  test("should show search component rendered", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />;
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test("should show input text with serch term and image of hero", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />;
      </MemoryRouter>
    );

    const inputText = screen.getByRole("textbox");
    expect(inputText.value).toBe("batman");

    const img = screen.getByRole("img");
    expect(img.src).toContain("http://localhost/assets/heroes/dc-batman.jpg");
  });

  test("should show alert of search hero", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />;
      </MemoryRouter>
    );

    const alertDoSearch = screen.getByLabelText("show-search");
    expect(alertDoSearch.style).toBeTruthy();
  });

  test("should show alert of error if don't find hero (vatnan) ", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=vatnan"]}>
        <SearchPage />;
      </MemoryRouter>
    );

    const alertErrorNotFinded = screen.getByLabelText("show-error");
    expect(alertErrorNotFinded.style.display).toBe("");
  });

  test("should call navigate with submit form", () => {
    const superhero = "superman";
    render(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchPage />;
      </MemoryRouter>
    );

    const inputValue = screen.getByRole("textbox");
    fireEvent.change(inputValue, { target: { value: superhero } });

    const form = screen.getByLabelText("form");
    fireEvent.submit(form);
    expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${superhero}`);
  });
});
