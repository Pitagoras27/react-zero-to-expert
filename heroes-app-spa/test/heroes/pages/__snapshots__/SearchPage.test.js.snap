// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test over search component should show search component rendered 1`] = `
<div>
  <div
    class="container m-3"
  >
    <div
      class="row"
    >
      <div
        class="col-md-5"
      >
        <h4>
          Search hero
        </h4>
        <hr />
        <form
          aria-label="form"
        >
          <input
            autocomplete="off"
            class="form-control"
            name="searchText"
            placeholder="search a hero"
            type="text"
            value="batman"
          />
          <button
            class="btn btn-outline-primary mt-3"
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
      <div
        class="col-7"
      >
        <h4>
          Results
        </h4>
        <hr />
        <div
          aria-label="show-search"
          class="alert alert-primary animate__animated animate__fadeIn"
          style="display: none;"
        >
          Search a hero
        </div>
        <div
          aria-label="show-error"
          class="alert alert-danger animate__animated animate__fadeIn"
          style="display: none;"
        >
          No hero with 
          <b>
            batman
          </b>
        </div>
        <div
          class="col animate__animated animate__fadeInUp"
        >
          <div
            class="card"
          >
            <div
              class="row no-glutters"
            >
              <div
                class="col-4"
              >
                <img
                  alt="Batman"
                  class="card-img"
                  src="assets/heroes/dc-batman.jpg"
                />
              </div>
              <div
                class="col-8"
              >
                <div
                  class="card-body"
                >
                  <h5
                    class="card-title"
                  >
                    Batman
                  </h5>
                  <p
                    class="card-text"
                  >
                    Bruce Wayne
                  </p>
                </div>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                >
                  MÃ¡s...
                </a>
              </div>
            </div>
            Batman
          </div>
        </div>
      </div>
    </div>
  </div>
  ;
</div>
`;
